File: go/binary_search/binary_search_find_last.go
package binary_search

// findLastInt should return last index of val in a sorted array arr, or -1 if not found
func findLastInt(arr []int, val int) int {
	l, r := 0, len(arr)

	for l < r {
		m := (l + r) / 2

		if arr[m] <= val {
			l = m + 1
		} else {
			r = m
		}
	}

	if r > 0 && arr[r-1] == val {
		return r - 1
	}

	return -1
}

File: java/src/binary_search/BinarySearchFindLast.java
package binary_search;

public class BinarySearchFindLast {
	// findLastInt should return last index of val in a sorted array arr, or -1 if not found
	public static int findLastInt(int[] arr, int val) {
		int l = 0, r = arr.length;

		while (l < r) {
			int m = (l + r)/2;
			if (arr[m] > val) {
				r = m;
			} else {
				l = m + 1;
			}
		}

		return l > 0 && arr[l - 1] ==val ? l - 1 : -1;
	}
}
